tf2_sensor_msgs/tf2_sensor_msgs/tf2_sensor_msgs.py

```diff
@@ -0,0 +1,142 @@
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+import numpy as np
+import numpy.typing as npt
+import tf2_ros
+from geometry_msgs.msg import Transform, TransformStamped
+from sensor_msgs.msg import PointCloud2
+from sensor_msgs_py.point_cloud2 import create_cloud, read_points
+from typing import Union
+
+
+def to_msg_msg(msg):
+    return msg
+
+
+tf2_ros.ConvertRegistration().add_to_msg(PointCloud2, to_msg_msg)
+
+
+def from_msg_msg(msg):
+    return msg
+
+
+tf2_ros.ConvertRegistration().add_from_msg(PointCloud2, from_msg_msg)
+
+
+def transform_points(
+        point_cloud: npt.ArrayLike,
+        transform: Union[Transform, TransformStamped]) -> np.ndarray:
+    """
+    Transform a bulk of points from an numpy array using a provided `Transform`.
+
+    :param point_cloud: nx3 Array of points where n is the number of points
+    :param transform: TF2 transform used for the transformation
+    :returns: Array with the same shape as the input array, but with the transformation applied
+    """
+    # Check if we have a TransformStamped
+    if isinstance(transform, TransformStamped):
+        transform = transform.transform
+
+    # Cast ArrayLike object to np.ndarray
+    point_cloud = np.asarray(point_cloud)
+
+    # Build affine transformation
+    transform_translation = np.array([
+        transform.translation.x,
+        transform.translation.y,
+        transform.translation.z
+    ])
+    transform_rotation_matrix = _get_mat_from_quat(
+        np.array([
+            transform.rotation.w,
+            transform.rotation.x,
+            transform.rotation.y,
+            transform.rotation.z
+        ]))
+
+    # "Batched" matmul meaning a matmul for each point
+    # First we offset all points by the translation part
+    # followed by a rotation using the rotation matrix
+    return np.einsum(
+        'ij, pj -> pi',
+        transform_rotation_matrix,
+        point_cloud) + transform_translation
+
+
+def do_transform_cloud(
+        cloud: PointCloud2,
+        transform: Union[Transform, TransformStamped]) -> PointCloud2:
+    """
+    Apply a `Transform` or `TransformStamped` on a `PointCloud2`.
+
+    The x, y, and z values are transformed into a different frame,
+    while the rest of the cloud is kept untouched.
+
+    :param cloud: The point cloud that should be transformed
+    :param transform: The transform which will applied to the point cloud
+    :returns: The transformed point cloud
+    """
+    points = list(read_points(  # TODO remove list and use numpy method
+        cloud,
+        field_names=["x", "y", "z"]))
+    points = transform_points(points, transform)  # TODO set points[["x", "y", "z"]]
+    # TODO add points to the rest of the points
```

Comment:
I plan to resolve them for this pr. But they needed the changes in https://github.com/ros2/common_interfaces/pull/175, and I was busy since it was merged.

Reply:
I resolved the last TODOs